### PR TITLE
ARROW-14095: [C++] subtract(timestamp, duration) -> timestamp kernel

### PR DESCRIPTION
**Pull request by rok**
[ARROW-14095](https://issues.apache.org/jira/browse/ARROW-14095)

### CODE REVIEW

**Comment by rok** [comment]
@lidavidm similar question about units as in #12124 - should we only cover matching time units (e.g. `subtract(timestamp[s], duration[s]) -> timestamp[s]`) or allow non-matching units too (e.g. `subtract(timestamp[s], duration[ms]) -> timestamp[s]`).

**Comment by jorisvandenbossche** [comment]
I would expect that for non-matching resolution in the input, we cast to the most detailed resolution. But, this doesn't necessarily need to have kernels for all combinations of resolutions. I think the casting step before calling the kernel should take care of that (like `add(int64, int32) -> int64` is not a registered kernel (only `add(int64, int64)` is), but still works).

The same is true for `subtract(timestamp, timestamp) -> duration`, which is for now only implemented for matching resolutions, and also the automatic casting doesn't seem to work yet for timestamps.

**Comment by rok** [comment]
@jorisvandenbossche heterogeneous time resolutions are now supported.

**Comment by rok** [comment]
@jorisvandenbossche It would appear test + doc only yes. [ARROW-14095](https://github.com/apache/arrow/pull/12136/files#diff-3eafd7246f6a8c699f10d46e3276852fe44b6853b5517ef10396e561730c09f4) added the `subtract(timestamp, duration)` kernel and dispatch best path to get from `subtract(date, duration)` -> `subtract(timestamp, duration)`.

File: docs/source/cpp/compute.rst [hunk 1]
```diff
@@ -450,9 +450,9 @@ Mixed time resolution temporal inputs will be cast to finest input resolution.
 +------------------+--------+----------------------------+----------------------------+-------+
 | sign             | Unary  | Numeric                    | Int8/Float32/Float64       | \(2)  |
 +------------------+--------+----------------------------+----------------------------+-------+
-| subtract         | Binary | Numeric/Date/Duration      | Numeric/Date/Duration      | \(1)  |
+| subtract         | Binary | Numeric/Timestamp/Duration | Numeric/Timestamp/Duration | \(1)  |
```
Comment (lidavidm) [inline docs/source/cpp/compute.rst:453]: Temporal instead of Timestamp?

File: cpp/src/arrow/compute/kernels/scalar_temporal_test.cc [hunk 2]
```diff
@@ -1030,6 +1030,80 @@ TEST_F(ScalarTemporalTest, TestTemporalSubtractDateAndDurationChecked) {
...
+                      ArrayFromJSON(timestamp_unit_ns, times_seconds_precision));
+  }
+
+  auto seconds_3 = ArrayFromJSON(timestamp(TimeUnit::SECOND), R"([3, null])");
+  auto milliseconds_2k = ArrayFromJSON(duration(TimeUnit::MILLI), R"([2000, null])");
+  auto milliseconds_1k = ArrayFromJSON(timestamp(TimeUnit::MILLI), R"([1000, null])");
+  CheckScalarBinary(op, seconds_3, milliseconds_2k, milliseconds_1k);
+
+  auto seconds_3_tz = ArrayFromJSON(timestamp(TimeUnit::SECOND, "UTC"), R"([3, null])");
+  auto milliseconds_1k_tz =
```
Comment (lidavidm) [inline cpp/src/arrow/compute/kernels/scalar_temporal_test.cc:1065]: It seems the _tz variants aren't used here

File: cpp/src/arrow/compute/kernels/scalar_temporal_test.cc [hunk 3]
```diff
@@ -1030,6 +1030,80 @@ TEST_F(ScalarTemporalTest, TestTemporalSubtractDateAndDurationChecked) {
...
+      ArrayFromJSON(timestamp(TimeUnit::MILLI, "UTC"), R"([1000, null])");
+  CheckScalarBinary(op, seconds_3, milliseconds_2k, milliseconds_1k);
```
Comment (lidavidm) [inline cpp/src/arrow/compute/kernels/scalar_temporal_test.cc:1104]: nor here

File: docs/source/cpp/compute.rst (hunk 1 above)
Comment (rok) [reply docs/source/cpp/compute.rst]: Changed `Numeric/Timestamp/Duration` -> ` Numeric/Temporal`

File: cpp/src/arrow/compute/kernels/scalar_temporal_test.cc (hunk 2 above)
Comment (rok) [reply cpp/src/arrow/compute/kernels/scalar_temporal_test.cc:1065]: Nice catch! Fixed.

File: cpp/src/arrow/compute/kernels/scalar_temporal_test.cc (hunk 3 above)
Comment (rok) [reply cpp/src/arrow/compute/kernels/scalar_temporal_test.cc]: Done.

**Review by lidavidm — APPROVED** [review]

**Pull request closed without merging**